gpushift/sklearn/meanshift.py

```diff
@@ -0,0 +1,321 @@
+        self.clustering_step = ClusteringStep(bandwidth, self._get_kernel(kernel), 
+                self._get_distance_metric(distance_metric), 
+                max_clusters=max_clusters, use_keops=False)
+
+    def _get_kernel(self, kernel_name):
+        r"""
+        Dispatch kernel functions.
+
+        Inputs:
+            :param kernel_name: str
+
+        Outputs:
+            unary callable operation, is a kernel function.
+        """
+        _kernels = {
+                'gaussian'  :       lambda x : ( -x / (2*self.bandwidth**2)).exp(),
+                'epanechnikov'  :   lambda x : 1 - (x**2)/(self.bandwidth**2),
+                'flat'  :           lambda x : x < self.bandwidth
+        }
+
+        return _kernels[kernel_name]
+
+    def _get_distance_metric(self, distance_metric_name):
+        r"""
+        Dispatch distance metric functions.
+
+        Inputs:
+            :param distance_metric_name: str, descriptive name of the kernel.
+
+        Outputs:
+            binary callable operation, its inputs are (n_samples, n_features),
+            and (m_samples, n_features). The output shape is (n_samples, 
+            m_samples).
+        """
+        _distance_metrics = {
+                'euclidean'  :  lambda x,y : (x[:,None,:]-y[None,:,:]).square().sum(-1).sqrt(),
+                'spherical'  :  lambda x,y : 1 - (x@y.T)**2,
+                'composite'  :  None
+        }
+
+        # Define separately to avoid huge lambda function
+        def _composite(x,y):
+            r"""
+            Assumed that the last two dimensions are flat, and the rest
+            are spherical.
+            """
+            euclid_ = _distance_metrics['euclidean']
+            sphere_ = _distance_metrics['spherical']
+            return euclid_(x[:,-2:],y[:,-2:])**2 + sphere_(x[:,:-2],y[:,:-2])**2
+
+        _distance_metrics['composite'] = _composite
+
+        return _distance_metrics[distance_metric_name]
+
+#    def get_sp_composite_distance_metric(self, dims=(16,2)):
+#        r"""
+#        The composite metric for self-parent segmentation and tracking
+#        algorithm. The space it works on is partly spherical and partly
+#        Euclidean, in that order of dimensions.
+#
+#        Inputs:
+#            :param dims: tuple (2,) ints. Shows the number of dimensions part
+#                of the spherical and Euclidean space, in that order. Default
+#                is (16,2) which means that the first 16 dimensions of the 
+#                vectors worked on, form a spherical space, and the latter 2
+#                are Euclidean.
+#
+#        Outputs:
+#            callable binary operation. The distance between vectors belonging
+#            to the composite vector space.
+#        """
+#        euclidean = self._get_distance_metric('euclidean')
+#        spherical = self._get_distance_metric('spherical')
+#
+#        composite = lambda x,y : euclidean(x,y)**2 + spherical(x,y)
+#
```

Comment:
If this block is no longer needed (as you have implemented differently above), I would prefer to delete it. Same for the other commented blocks above.